src/webgpu/shader/execution/expression/call/builtin/subgroupAdd.spec.ts

```diff
@@ -0,0 +1,546 @@
+ * @param values An of two values containing the interesting values in the input
+ */
+function checkAccuracy(
+  metadata: Uint32Array,
+  output: Float32Array | Float16Array,
+  indices: number[],
+  values: number[],
+  type: 'f16' | 'f32'
+): undefined | Error {
+  const subgroupIdIdx1 = metadata[1 + indices[0]];
+  const subgroupIdIdx2 = metadata[1 + indices[1]];
+  for (let i = 0; i < output.length; i++) {
+    const subgroupId = metadata[1 + i];
+
+    const v1 = subgroupId === subgroupIdIdx1 ? values[0] : kIdentity;
+    const v2 = subgroupId === subgroupIdIdx2 ? values[1] : kIdentity;
+    const interval =
+      type === 'f16' ? FP.f16.additionInterval(v1, v2) : FP.f32.additionInterval(v1, v2);
+    if (!interval.contains(output[i])) {
+      return new Error(`Invocation ${i}, subgroup id ${subgroupId}: incorrect result
+- interval: ${interval.toString()}
+- output: ${output[i]}`);
+    }
+  }
+
+  return undefined;
+}
+
+g.test('fp_accuracy')
+  .desc(
+    `Tests the accuracy of floating-point addition.
+
+The order of operations is implementation defined, most threads are filled with
+the identity value and two receive random values.
+Subgroup sizes are not known ahead of time so some cases may not perform any
+interesting operations. The test biases towards checking subgroup sizes under 64.`
+  )
+  .params(u =>
+    u
+      .combine('case', [...iterRange(kNumCases, x => x)])
+      .combine('type', ['f32', 'f16'] as const)
+      .combine('wgSize', [
+        [kStride, 1, 1],
+        [kStride / 2, 2, 1],
+      ] as const)
+  )
+  .beforeAllSubcases(t => {
+    const features: GPUFeatureName[] = ['subgroups' as GPUFeatureName];
+    if (t.params.type === 'f16') {
+      features.push('shader-f16');
+      features.push('subgroups-f16' as GPUFeatureName);
+    }
+    t.selectDeviceOrSkipTestCase(features);
+  })
+  .fn(async t => {
+    const prng = new PRNG(t.params.case);
+
+    // Compatibility mode has lower workgroup limits.
+    const wgThreads = t.params.wgSize[0] * t.params.wgSize[1] * t.params.wgSize[2];
+    const {
+      maxComputeInvocationsPerWorkgroup,
+      maxComputeWorkgroupSizeX,
+      maxComputeWorkgroupSizeY,
+      maxComputeWorkgroupSizeZ,
+    } = t.device.limits;
+    t.skipIf(
+      maxComputeInvocationsPerWorkgroup < wgThreads ||
+        maxComputeWorkgroupSizeX < t.params.wgSize[0] ||
+        maxComputeWorkgroupSizeY < t.params.wgSize[1] ||
+        maxComputeWorkgroupSizeZ < t.params.wgSize[2],
+      'Workgroup size too large'
+    );
+
+    // Bias half the cases to lower indices since most subgroup sizes are <= 64.
+    let indexLimit = kStride;
+    if (t.params.case < kNumCases / 4) {
+      indexLimit = 16;
+    } else if (t.params.case < kNumCases / 2) {
+      indexLimit = 64;
+    }
+
+    // Ensure two distinct indices are picked.
+    const idx1 = prng.uniformInt(indexLimit);
+    let idx2 = prng.uniformInt(indexLimit - 1);
```

Comment:
This code sequence cleverly ensures idx2 < indexLimit. Might be worth an assert?